### PR TITLE
Add --open flag to `cabal haddock`

### LINKED ISSUE #7366: Support `--open` option for `cabal haddock`
It would be nice to have `cabal haddock --open` to automatically open the generated Haddock documentation (more specifically, the `index.html`) in a browser, similar to `stack haddock --open` for Stack. If this functionality already exists (I didn't find anything), it would be nice to alias it to the `--open` option.

### PR DESCRIPTION
**Pull request by patrickt**
Both `stack` and other package managers like `cargo` have an `--open`
flag for the common case of opening documentation locally for
inspection or reference. This patch adds a `--open` flag, which when set
tries to open the HTML file associated with the generated
documentation, either through `xdg-open` (Linux) or `open` (macOS).

This is my first patch to `cabal`, so please inform me if I’ve made some obvious error apparent to someone with more experience in this codebase. I haven’t yet ported this functionality to `haddock` itself, but I can should that be necessary.

I tested this by, in the `cabal` project, invoking `cabal run cabal-install:exe:cabal -- haddock Cabal --open`.

Fixes #7366.

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

### CODE REVIEW

**Review by emilypi — APPROVED** [review]
👍 for quality of life improvements

**Comment by emilypi** [comment]
Looks like there are some package tests that are failing. When CI's green, I'll merge

**Comment by gbaz** [comment]
related: https://github.com/haskell/cabal/issues/5803

**Comment by gbaz** [comment]
btw is there anything appropriate to call on Windows?

Also, not to bikeshed or goldplate too much, but it might be nice to have an option to just output the index location without invoking anything on it, for ide usage or the like, as discussed in the above linked ticket.

(nb: I consider neither of these considerations a blocker, but rather something that could be addressed either in this or in future PRs).

File: Cabal/src/Distribution/Simple/Haddock.hs
Comment (fgaz) [inline Cabal/src/Distribution/Simple/Haddock.hs:577]: * I think this should depend on the platform.
* I'm not sure whether this should use the whole `Distribution/Simple.Program` thing, but it's something to consider. It looks like no other place uses `findExecutable` directly, relying on `ProgramDB instead`. On the other hand, this is a `cabal-install` exclusive command, and I don't know if there's a way to keep the `ProgramDB` changes out of Cabal.

File: changelog.d/pr-7550
Comment (fgaz) [inline changelog.d/pr-7550:8]: since it's a single item it can go in the synopsis instead, otherwise on the changelog it will appear as something like:

> * Add --open flag to `cabal haddock`
>   * Add --open flag to `haddock` subcommand that attempts to open the generated documentation in a browser.

**Comment by patrickt** [comment]
@emilypi Yeah, the QuickCheck tests to convert between legacy and new-style project configs are failing (specifically the ProjectConfig printing/parsing round trip). I can fix them, but wanted to know if you think I’m going about this the right way in terms of plumbing these options through to the various different record types.

File: Cabal/src/Distribution/Simple/Haddock.hs
Comment (patrickt) [reply Cabal/src/Distribution/Simple/Haddock.hs:577]: Re. platform: I agree. It looks like I can check the `LocalBuildInfo` for the current platform, so that should work. I’ll dispatch to `xdg-open` on Linux and the BSDs. @gbaz pointed out that I didn’t account for a Windows case; [these docs](https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/start) say that `start` should work.

File: Cabal/src/Distribution/Simple/Haddock.hs
Comment (fgaz) [inline Cabal/src/Distribution/Simple/Haddock.hs:581]: I think the error should be treated fatally, since here the command is failing to do something the user requested

File: Cabal/src/Distribution/Simple/Haddock.hs
Comment (fgaz) [reply Cabal/src/Distribution/Simple/Haddock.hs]: Good point. I'd still move all the open-file logic in a separate module, since it's independent from haddock stuff

**Review by fgaz — COMMENTED** [review]
Should this really go in Cabal? It's a low level interface with which one does not usually interact directly, so maybe cabal-install could be a better place to implement this. @emilypi @gbaz what do you think?

**Comment by gbaz** [comment]
That's a good point. But also I'm not sure if the invocation of `Cabal`'s haddock passes back enough info so that `open` can be moved to `cabal-install`? If so, that's better practice -- we want to keep Cabal-the-library as frill-free as possible, imho.

**Comment by patrickt** [comment]
@fgaz @gbaz The problem I ran into when trying to keep this in `cabal-install` is that the information about where the documentation is written is contained entirely inside `Cabal` (specifically in the `renderArgs` call in `Distribution.Simple.Haddock`). It’s not clear to me where I would access this information in, say, `Distribution.Client.CmdHaddock.haddockAction`, though I might be missing something.

File: Cabal/src/Distribution/Simple/Haddock.hs
Comment (patrickt) [reply Cabal/src/Distribution/Simple/Haddock.hs]: Moved the code that cases over the `OS` to `Distribution.System`, where it can live alongside the `Platform` type.

File: Cabal/src/Distribution/Simple/Haddock.hs
Comment (patrickt) [reply Cabal/src/Distribution/Simple/Haddock.hs:581]: Changed that `warn` to `die’`.

**Comment by gbaz** [comment]
I _think_ this would be the place to modify if you wanted to keep it in cabal-install...

https://github.com/haskell/cabal/blob/bd5e3f4f2e5be52d514af3a5353725cb69bd77eb/cabal-install/src/Distribution/Client/ProjectBuilding.hs#L1312

**Comment by patrickt** [comment]
@fgaz @gbaz I’ve moved all the code to cabal-install, and now like this patch a lot more: thank you for all your help! This also gives us the opportunity to choose either `--haddock-open` or `--open` for the flag; does anyone have a strong preference?

**Comment by fgaz** [comment]
You can move the flag definition to cabal-install too. You could create a `data TestFlags` in `Distribution.Client.Setup` like most other command flags or in its own file like `ClientInstallFlags` (by the way we should probably move one or the other in a single place for consistency)

**Comment by fgaz** [comment]
> This also gives us the opportunity to choose either --haddock-open or --open for the flag; does anyone have a strong preference?

weak preference for --open: i doubt we'll have a global flag with the same name. it does conflict with all the --haddock-* flags too

**Comment by patrickt** [comment]
I too prefer `--open`. Will make the change.

**Comment by patrickt** [comment]
@fgaz Re. moving the flag definition: if I created a new `HaddockFlags` type in `cabal-install`, would I just port over all the associated fields and convert between them at interface boundaries, or is there a more elegant way to do this?

**Comment by fgaz** [comment]
> port over all the associated fields

no, ideally the existing fields in Cabal should remain in Cabal, without copying. the new `HaddockFlags` should only contain cabal-install-specific flags (maybe it could be called `ClientHaddockFlags` or `HaddockExFlags`, mirroring `ClientInstallFlags`/`ConfigExFlags`[1]), and the haddock command will use both, like the configure command does

[1]: here too we should settle on a convention :-/

**Comment by patrickt** [comment]
@fgaz Ah, I see, so `haddockCommand` will have type `CommandUI (Cabal.HaddockFlags, ClientHaddockFlags)`. Gotcha. I will take a run at it and see how far I get.

**Comment by patrickt** [comment]
This is ready for another look. Thanks to @fgaz’s advice and the fact that `NixStyleFlags` has an `extraFlags` field of customizable type, which before was just `()`, I can now stick a new `ClientHaddockFlags` type there and then thread that information through to the `BuildTimeSettings`, which is propagated appropriately to `buildInplaceUnpackedPackage`. So now there are no changes either to `Cabal` or the `cabal-install` unit tests! Way nicer.

**Comment by patrickt** [comment]
CI seems stuck, not sure why.

**Review by fgaz — COMMENTED** [review]
Looks good, thanks! A squash and I'd say it's ready to merge

**Comment by patrickt** [comment]
@fgaz Squashed! Good to go once you approve the PR & CI workflows and the build goes green.

**Review by fgaz — APPROVED** [review]
Thanks! (i force-pushed a fix for the wrong ticket number in the chagelog entry)

**Pull request merged**
